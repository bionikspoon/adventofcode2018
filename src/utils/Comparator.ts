type CompareFn<T> = (l: T, r: T) => 0 | -1 | 1

export default class Comparator<T> {
  private static defaultCompareFunction<T>(l: T, r: T) {
    if (l === r) return 0

    return l < r ? -1 : 1
  }
  private compare: CompareFn<T>
  constructor(fn?: CompareFn<T>) {
    this.compare = fn || Comparator.defaultCompareFunction
  }

  public equal(l: T, r: T) {
    return this.compare(l, r) === 0
  }
  public lessThan(l: T, r: T) {
    return this.compare(l, r) < 0
  }
  public greaterThan(l: T, r: T) {
    return this.compare(l, r) > 0
  }
  public lessThanOrEqual(l: T, r: T) {
    return this.lessThan(l, r) || this.equal(l, r)
  }
  public greaterThanOrEqual(l: T, r: T) {
    return this.greaterThan(l, r) || this.equal(l, r)
  }
  public reverse() {
    const compareOriginal = this.compare
    this.compare = (l, r) => compareOriginal(r, l)
    return this
  }
}
