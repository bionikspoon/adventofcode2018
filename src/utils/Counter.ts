import R from 'ramda'

export default class Counter {
  private counts: { [key: string]: number }

  constructor() {
    this.counts = {}
  }

  public add(value: string | number) {
    const count = this.counts[value] ? this.counts[value] + 1 : 1

    this.counts = { ...this.counts, [value]: count }
    return this
  }

  public deleteValue(v: number) {
    this.counts = R.reject(R.equals(v), this.counts)

    return this
  }

  public getValue(key: string | number) {
    if (!this.counts[key]) return 0

    return this.counts[key]
  }

  public mostCommon() {
    return this.entries().sort(
      ([lKey, lCount], [rKey, rCount]) => rCount - lCount
    )
  }

  public values() {
    return Object.values(this.counts)
  }

  public entries() {
    return Object.entries(this.counts)
  }
}
