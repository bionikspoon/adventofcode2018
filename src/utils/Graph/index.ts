// Borrowed from https://github.com/trekhleb/javascript-algorithms/tree/master/src/data-structures/graph
export { default as Graph } from './Graph'
export { default as GraphEdge } from './GraphEdge'
export { default as GraphVertex } from './GraphVertex'
