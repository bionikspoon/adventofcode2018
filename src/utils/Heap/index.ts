export { default as MinHeap } from './MinHeap'
