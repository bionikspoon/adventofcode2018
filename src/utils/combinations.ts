export default function combinations<T>(items: T[]): Array<[T, T]> {
  return Array.from(iterateCombinations(items))
}

function* iterateCombinations<T>(items: T[]): IterableIterator<[T, T]> {
  for (let i = 0; i < items.length; i++) {
    const I = items[i]
    for (let j = i + 1; j < items.length; j++) {
      const J = items[j]
      yield [I, J]
    }
  }
}
