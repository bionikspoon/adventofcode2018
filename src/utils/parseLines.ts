export default function parseLines(input: string) {
  return input
    .trim()
    .split('\n')
    .map(line => line.trim())
}
