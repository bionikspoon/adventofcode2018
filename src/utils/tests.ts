import fs from 'fs'
import path from 'path'
import { promisify } from 'util'

const readFileAsync = promisify(fs.readFile)

export const getInput = (dirname: string, file: string) =>
  readFileAsync(path.join(dirname, 'case', file)).then(buffer =>
    buffer.toString()
  )
