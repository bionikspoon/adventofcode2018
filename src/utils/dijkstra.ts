import { Graph, GraphVertex } from './Graph'
import PriorityQueue from './PriorityQueue'

interface IDistances {
  [key: string]: number
}
export interface IDijkstraOptions<T> {
  queueCompareFn?: (
    this: PriorityQueue<T>,
    l: GraphVertex<T>,
    r: GraphVertex<T>
  ) => 0 | -1 | 1
  canTraverse?: (neighbor: GraphVertex<T>) => boolean
  shouldPoll?: (vertex: GraphVertex<T>) => boolean
}
export default function dijkstra<T>(
  graph: Graph<T>,
  startVertex: GraphVertex<T>,
  options: IDijkstraOptions<T> = {}
) {
  const visitedVertices: { [key: string]: GraphVertex<T> } = {}
  const queue = new PriorityQueue<GraphVertex<T>>(options.queueCompareFn)
  const { distances, previousVertices } = initDistances(graph)

  distances[startVertex.getKey()] = 0
  queue.add(startVertex, distances[startVertex.getKey()])

  while (!queue.isEmpty()) {
    const currentVertex = queue.poll()!

    graph.getNeighbors(currentVertex).forEach(neighbor => {
      if (visitedVertices[neighbor.getKey()]) return
      if (options.canTraverse && !options.canTraverse(neighbor)) return

      const distanceToNeighborFromCurrent =
        distances[currentVertex.getKey()] +
        graph.findEdge(currentVertex, neighbor)!.weight

      if (distanceToNeighborFromCurrent < distances[neighbor.getKey()]) {
        distances[neighbor.getKey()] = distanceToNeighborFromCurrent
        updateNeighborPriority(queue, neighbor, distances)
        previousVertices[neighbor.getKey()] = currentVertex
      }

      ensureQueueHasNeighbor(queue, neighbor, distances)
    })

    visitedVertices[currentVertex.getKey()] = currentVertex
    if (options.shouldPoll && !options.shouldPoll(currentVertex)) break
  }

  return { distances, previousVertices }
}
function initDistances<T>(graph: Graph<T>) {
  const distances: IDistances = {}
  const previousVertices: { [key: string]: GraphVertex<T> | null } = {}

  graph.getAllVertices().forEach(vertex => {
    distances[vertex.getKey()] = Infinity
    previousVertices[vertex.getKey()] = null
  })

  return { distances, previousVertices }
}

function updateNeighborPriority<T>(
  queue: PriorityQueue<GraphVertex<T>>,
  neighbor: GraphVertex<T>,
  distances: IDistances
): void {
  if (queue.hasValue(neighbor)) {
    queue.changePriority(neighbor, distances[neighbor.getKey()])
  }
}

function ensureQueueHasNeighbor<T>(
  queue: PriorityQueue<GraphVertex<T>>,
  neighbor: GraphVertex<T>,
  distances: IDistances
): void {
  if (!queue.hasValue(neighbor)) {
    queue.add(neighbor, distances[neighbor.getKey()])
  }
}
