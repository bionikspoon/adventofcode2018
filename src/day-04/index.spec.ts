import moment from 'moment'
import util from 'util'
import {
  EntryType,
  findSleepingGuard,
  sortParsedRecords,
  toDate,
  toGuardRecords,
  toParsedEntry,
} from '.'
import parseLines from '../utils/parseLines'
import { getInput } from '../utils/tests'

moment.prototype[util.inspect.custom] = function inspect() {
  return this.inspect()
}

describe('part 1 - #findSleepingGuard', () => {
  test.each`
    file                   | expected
    ${'part-1-case-1.txt'} | ${240}
    ${'input.txt'}         | ${106710}
  `('it finds the overlapping claims in $file', async ({ file, expected }) => {
    const input = await getInput(__dirname, file)

    expect(findSleepingGuard(input)).toEqual(expected)
  })
})

describe('#toParsedEntry', () => {
  test.each`
    line                                           | id           | date                          | note
    ${'[1518-11-01 00:00] Guard #10 begins shift'} | ${10}        | ${toDate('1518-11-01 00:00')} | ${'begins shift'}
    ${'[1518-11-01 00:05] falls asleep'}           | ${undefined} | ${toDate('1518-11-01 00:05')} | ${'falls asleep'}
    ${'[1518-11-01 00:25] wakes up'}               | ${undefined} | ${toDate('1518-11-01 00:25')} | ${'wakes up'}
  `('it parses string into an object', ({ line, id, date, note }) => {
    expect(toParsedEntry(line)).toEqual({
      id,
      note,
      date,
    })
  })
})

describe('#toGuardRecord', () => {
  let parsedEntries: EntryType[]

  describe('given part-1-case-1.txt', () => {
    beforeEach(async () => {
      const input = await getInput(__dirname, 'part-1-case-1.txt')
      parsedEntries = sortParsedRecords(parseLines(input))
    })

    test('it matches snapshot', () => {
      expect(parsedEntries).toMatchSnapshot()
    })

    test('it converts entries to records', () => {
      const subject = toGuardRecords(parsedEntries).map((entry: any) =>
        entry.toString()
      )

      const expected = [
        '11-01  #0010  .....####################.....#########################.....',
        '11-02  #0099  ........................................##########..........',
        '11-03  #0010  ........................#####...............................',
        '11-04  #0099  ....................................##########..............',
        '11-05  #0099  .............................................##########.....',
      ]

      expect(subject).toEqual(expected)
    })
  })
  describe('given input.txt', () => {
    beforeEach(async () => {
      const input = await getInput(__dirname, 'input.txt')

      parsedEntries = sortParsedRecords(parseLines(input))
    })

    test('it matches snapshot', () => {
      expect(parsedEntries).toMatchSnapshot()
    })
  })
})
