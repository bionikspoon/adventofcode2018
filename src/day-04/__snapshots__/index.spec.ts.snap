// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#toGuardRecord given "input.txt" it matches snapshot 1`] = `
Array [
  "03-07  #1291  ...................########.................................",
  "03-08  #0773  .........####################################...............",
  "03-09  #0863  .................................########...................",
  "03-10  #3191  .......................................................#....",
  "03-11  #2791  ............................................##############..",
  "03-12  #2749  ................###############...............#.............",
  "03-13  #2539  ............................#################...............",
  "03-14  #3257  ..........................#############.........###########.",
  "03-15  #0269  ......................#######.........##################....",
  "03-16  #3557  ######################.................###################..",
  "03-17  #1237  ..........................###########....##############.....",
  "03-18  #0269  ..........................#########################.........",
  "03-19  #2251  ......................................................###...",
  "03-20  #0863  ...........####################...........#############.....",
  "03-21  #1291  ............................####........#......####.........",
  "03-22  #0631  ...######################################################...",
  "03-23  #1367  .........###############################################....",
  "03-24  #0631  ......########################################..............",
  "03-25  #1367  ...#######################################..................",
  "03-26  #0269  ####################################....##..................",
  "03-27  #1367  ..............##########........######################......",
  "03-28  #1367  ......##########............###############################.",
  "03-29  #0773  ............####################################....###.....",
  "03-30  #0677  ..........................#############.....................",
  "03-31  #1291  .##################################################.........",
  "04-01  #2749  ........................################################....",
  "04-02  #2749  ....................###############################.........",
  "04-03  #3557  ..............##########################################....",
  "04-04  #0433  ............................................................",
  "04-05  #3041  .............#########################################......",
  "04-06  #1367  ...................###############..........................",
  "04-07  #0773  ################################################............",
  "04-08  #3191  ...############.............................................",
  "04-09  #0269  ..............##########################...##...............",
  "04-10  #3257  .........#########################.................######...",
  "04-11  #0677  ..................##########################................",
  "04-12  #0449  .............#########################################......",
  "04-13  #2539  ............#######...#########################......#####..",
  "04-14  #0677  ...........###########################################......",
  "04-15  #1237  ...................############################.............",
  "04-16  #1237  ........################################........########....",
  "04-17  #0269  ..................................###....##############.....",
  "04-18  #0863  ........................##################################..",
  "04-19  #1873  ..#########.......#################..................######.",
  "04-20  #2749  .........................##############.....................",
  "04-21  #2539  .....................##############################.........",
  "04-22  #0677  .........................####################...............",
  "04-23  #1367  .......................#####................................",
  "04-24  #1237  ..........###########################.......########........",
  "04-25  #3041  ..........................................################..",
  "04-26  #2539  ...............#############################................",
  "04-27  #1237  ................................######################......",
  "04-28  #2251  ....................###################################.....",
  "04-29  #1367  ........................#######...#########......#########..",
  "04-30  #0433  ............................................................",
  "05-01  #3041  .........................####################....##.........",
  "05-02  #1291  ......................#########################.............",
  "05-03  #2699  ..................................................######....",
  "05-04  #1873  .....................##.....##################..............",
  "05-05  #0631  ............###...................##......###...............",
  "05-06  #1873  ............................................#############...",
  "05-07  #3257  ..........................#########################.........",
  "05-08  #0269  .......................###################..................",
  "05-09  #3041  .......................................#####.............#..",
  "05-10  #3257  ..............................................##########....",
  "05-11  #3557  .............................##############.................",
  "05-12  #3557  ..............................################..............",
  "05-13  #0677  ###################################...............#.........",
  "05-14  #3257  ...................................##############...........",
  "05-15  #0863  ................########................................##..",
  "05-16  #1291  ............................###########################.....",
  "05-17  #1237  .......................#####################......#########.",
  "05-18  #1171  ...............#########################....................",
  "05-19  #0269  .....................................#####........######....",
  "05-20  #2749  ................################.......................###..",
  "05-21  #1237  .................................#########.........########.",
  "05-22  #0863  ..............##########################################....",
  "05-23  #0863  ......................................#...........#########.",
  "05-24  #3557  ##########################################...............##.",
  "05-25  #2791  ....###########################..........#########.......#..",
  "05-26  #3041  .....................######................################.",
  "05-27  #1291  .........########################################......##...",
  "05-28  #2749  ....#######################################################.",
  "05-29  #0631  .....................###############........................",
  "05-30  #1171  ................................##########################..",
  "05-31  #1291  ......................###########.......#...................",
  "06-01  #1873  ..................................................##........",
  "06-02  #0631  ...................................##########...............",
  "06-03  #3041  ..........###############################################...",
  "06-04  #0863  ................#########################################...",
  "06-05  #0449  .......................############################.........",
  "06-06  #2539  .....##...#########################################.........",
  "06-07  #3557  ...........###...############################...............",
  "06-08  #3557  ....################################........................",
  "06-09  #0677  ........................#########...........................",
  "06-10  #2251  .......................................##################...",
  "06-11  #0449  ......................#################################.....",
  "06-12  #3019  ............................................................",
  "06-13  #0269  .......................................#####........###.....",
  "06-14  #3557  ..............######################################........",
  "06-15  #3557  ...............##########################..........########.",
  "06-16  #2699  ............############..........########################..",
  "06-17  #2791  .......................####........#####.......###########..",
  "06-18  #0269  ...................................#####................#...",
  "06-19  #1171  .......................################################.....",
  "06-20  #3191  .............................................####........##.",
  "06-21  #0449  ..#####################.....................................",
  "06-22  #3557  .......###################################..................",
  "06-23  #3557  ........#####################...............................",
  "06-24  #1367  ..........#############################################.....",
  "06-25  #3257  ............................##############################..",
  "06-26  #3041  ..........######################.....#########......##......",
  "06-27  #3257  .......................................###.......##.....###.",
  "06-28  #3257  ............................##########################......",
  "06-29  #1367  #######################################.........##########..",
  "06-30  #2539  ..........#####............................#####...#....##..",
  "07-01  #1291  ..................................##################........",
  "07-02  #3041  .................................#########..................",
  "07-03  #2539  ...............................##########...................",
  "07-04  #0449  .......................#######..............................",
  "07-05  #1171  ......###..........................##.................##....",
  "07-06  #0449  .....................####...................................",
  "07-07  #3557  ...................###################.....####.............",
  "07-08  #2699  ............................###############################.",
  "07-09  #3257  ........######################################..............",
  "07-10  #3257  ......................................................#####.",
  "07-11  #3191  .............................#########################......",
  "07-12  #1171  .................................................###........",
  "07-13  #2699  #####################........................##########.....",
  "07-14  #3041  ................#######################...#############.....",
  "07-15  #0677  ....................................................###.....",
  "07-16  #1237  ................######################......................",
  "07-17  #3041  ..........................#######################...........",
  "07-18  #1291  .................................##########################.",
  "07-19  #1873  ..........................####.............####.............",
  "07-20  #1237  .........................................#.....###...####...",
  "07-21  #1873  ..................######........................#...........",
  "07-22  #0677  ....####################################....................",
  "07-23  #3257  .....................############################...........",
  "07-24  #1237  #######################################################.....",
  "07-25  #2251  ......................#####################..........###....",
  "07-26  #0863  ...###########################.....................#######..",
  "07-27  #3191  ......................................###################...",
  "07-28  #0631  .................................#######################....",
  "07-29  #0631  .......###########################......#########...........",
  "07-30  #2749  .....................######################........##....##.",
  "07-31  #0863  ............................#######.........##########......",
  "08-01  #1171  ...###########################################........####..",
  "08-02  #1367  ............#################################...............",
  "08-03  #1171  .......................####################....#########....",
  "08-04  #0449  .............###############................................",
  "08-05  #1171  ........................################################....",
  "08-06  #2699  ......................###...................................",
  "08-07  #3041  ......#############################################.........",
  "08-08  #2699  ............#####################################....#####..",
  "08-09  #3557  ..................................###.......................",
  "08-10  #1873  .........................#########....#####################.",
  "08-11  #3191  ...........................................#############....",
  "08-12  #2791  ..##############............................................",
  "08-13  #0449  .........#########################################..........",
  "08-14  #0269  ......................###########......####################.",
  "08-15  #3257  ...................................................#######..",
  "08-16  #1291  ..........#####################..........#.....#####........",
  "08-17  #0269  ###########################...############..................",
  "08-18  #0269  .......................................#############........",
  "08-19  #0449  ...................##########################...............",
  "08-20  #1873  ..........################################################..",
  "08-21  #2539  .................############...............................",
  "08-22  #3019  ............................................................",
  "08-23  #2749  .#####################.....###......#.......................",
  "08-24  #2749  ....#####################################################...",
  "08-25  #0773  .........................################.......####........",
  "08-26  #2539  ....................................######..................",
  "08-27  #3041  .................#.......########.........##########........",
  "08-28  #1873  .........................................................##.",
  "08-29  #2791  ...........................#################........#######.",
  "08-30  #3041  ...........##########################################.......",
  "08-31  #1237  ...................................#######################..",
  "09-01  #3041  ...####################################################.....",
  "09-02  #0677  .........................................#####..............",
  "09-03  #3257  ..........#########################.......##########........",
  "09-04  #2251  ...............#########...............############.........",
  "09-05  #2699  .....####################...................................",
  "09-06  #2539  ..........######################################............",
  "09-07  #0269  ...................#####################################....",
  "09-08  #2791  ............................######.................##.......",
  "09-09  #1237  ................#########################################...",
  "09-10  #2791  ......#########.........................................#...",
  "09-11  #0863  ...................................###########...####.......",
  "09-12  #3019  ............................................................",
  "09-13  #2791  .................###################............###########.",
  "09-14  #3257  .....................................#####################..",
  "09-15  #0677  ..........#############################....#######...###....",
  "09-16  #1171  .....................################################....#..",
  "09-17  #2699  #####################################.......................",
  "09-18  #0449  ####################################............######...##.",
  "09-19  #3557  ...........#####################################.....#####..",
  "09-20  #2749  ..........##...........................############.........",
  "09-21  #2749  .............#########################.......###########....",
  "09-22  #3557  ..........######################............................",
  "09-23  #2791  .........#################............................###...",
  "09-24  #0449  ...................#########################.......#........",
  "09-25  #1367  .###################################....#################...",
  "09-26  #2539  .......................................##########...........",
  "09-27  #0269  .....................####################...................",
  "09-28  #0449  #####################.......................................",
  "09-29  #2699  .............############################...................",
  "09-30  #3191  ........######################......####################....",
  "10-01  #3191  .........................................................##.",
  "10-02  #3191  .............................#############################..",
  "10-03  #3191  ..........................................####..............",
  "10-04  #3257  ....................................#########...............",
  "10-05  #2791  ..................................########################..",
  "10-06  #1367  ........##########################..........................",
  "10-07  #0269  ......................................#####.....######......",
  "10-08  #0773  ........................#######################.............",
  "10-09  #3557  ...............###################..........................",
  "10-10  #2749  ...................#######################..................",
  "10-11  #0863  ............................############################....",
  "10-12  #0269  .....###############################...#########............",
  "10-13  #0239  ............................................................",
  "10-14  #2539  ...............########################............#######..",
  "10-15  #0449  ...................###########################..............",
  "10-16  #3041  .#########....##.......##########.......###############.....",
  "10-17  #0773  ...................##.............................#####.....",
  "10-18  #0863  ...................................########################.",
  "10-19  #3557  ..............................#...........###......####.....",
  "10-20  #1171  ......................................####################..",
  "10-21  #2699  ........#########################.............############..",
  "10-22  #3257  ...............................######################.......",
  "10-23  #0269  ....................................#####.............####..",
  "10-24  #3557  ...........................##############............###....",
  "10-25  #3557  ............###########################################.....",
  "10-26  #1367  ...................................................##....#..",
  "10-27  #1367  ..............................#######........############...",
  "10-28  #3191  ....................................................####....",
  "10-29  #1367  ....############################....####################....",
  "10-30  #0677  .........................................########...........",
  "10-31  #0677  ....##########################...######.....................",
  "11-01  #2539  ........#####################################.....#######...",
  "11-02  #0269  ........................#####################...............",
  "11-03  #2251  ..................######...######..............#######......",
  "11-04  #1367  ......#########################.....###########.....#####...",
  "11-05  #3257  ...........######################################...........",
  "11-06  #0773  ..................#############.............#######.........",
  "11-07  #3257  ...........#########.....................##############.....",
  "11-08  #2749  ................................######################...##.",
  "11-09  #0449  ..........###########.......................................",
  "11-10  #2791  .........................#############......................",
  "11-11  #3191  .................................######.....##############..",
  "11-12  #2749  ##############################################..............",
  "11-13  #0269  ....###########........................#########............",
  "11-14  #0677  ....................................................#....##.",
  "11-15  #0631  ....#################################.......................",
  "11-16  #3041  ....................................###########.............",
  "11-17  #3041  .................########################...................",
  "11-18  #1291  ...................................#################........",
  "11-19  #0239  ............................................................",
  "11-20  #2749  .................###########..........#################.....",
  "11-21  #0449  ......##########################............................",
  "11-22  #2749  .....###########################.....############...........",
  "11-23  #1237  ..............#####################################.........",
]
`;

exports[`#toGuardRecord given "part-1-case-1.txt" it matches snapshot 1`] = `
Array [
  "11-01  #0010  .....####################.....#########################.....",
  "11-02  #0099  ........................................##########..........",
  "11-03  #0010  ........................#####...............................",
  "11-04  #0099  ....................................##########..............",
  "11-05  #0099  .............................................##########.....",
]
`;
